### DISCUSSION TITLE
count of ibd segments between all sampled pairs

### DISCUSSION THREAD

**Discussion by hancockzb** [tskit-dev/tskit#1241] [Q&A] [answered]
Hi all,

Quick and simple question that me, a python layman, am struggling with: I'd like to count the number of genomic segments in which individual i and individual j in range(number of individuals) have the same MRCA. Goal here is to compare number of ibd segments between modern-modern individuals sampled across space and modern-ancient individuals across space and time. 

Thanks so much for your time and all the hard work you all have put into this package! 

Best,
Zach

**Comment by jeromekelleher** [comment, marked as answer]
Hi Zach! :wave:

This is actually not a simple question at all - computing IBD segments efficiently is not easy, and we don't have any public APIs for doing this at the moment. Here's a simple brute-force approach though, that you might be able to adapt for your purposes:
```python

import itertools
import dataclasses
import msprime


@dataclasses.dataclass
class IbdSegment:
    pair: (int, int)
    mrca: int
    left: float = 0
    right: float = -1


ts = msprime.simulate(5, recombination_rate=0.01, length=40, random_seed=42)
print(ts.draw_text())

pairs = list(itertools.combinations(ts.samples(), 2))
tree = ts.first()
mrca_map = {pair: IbdSegment(pair, tree.mrca(*pair)) for pair in pairs}
ibd_segs = []

while tree.next():
    for pair in pairs:
        ibd = mrca_map[pair]
        mrca = tree.mrca(*pair)
        if mrca != ibd.mrca:
            ibd.right = tree.interval[0]
            ibd_segs.append(ibd)
            # Start a new record
            mrca_map[pair] = IbdSegment(pair, mrca, tree.interval[0])

# Flush the remaining IBD segments
for ibd in mrca_map.values():
    ibd.right = ts.sequence_length
    ibd_segs.append(ibd)

for seg in ibd_segs:
    print(seg)
```

This gives:
```
1.72┊     9     ┊       9   ┊  
    ┊  ┏━━┻━┓   ┊     ┏━┻━┓ ┊  
1.72┊  8    ┃   ┊     ┃   ┃ ┊  
    ┊ ┏┻┓   ┃   ┊     ┃   ┃ ┊  
0.87┊ ┃ ┃   7   ┊     7   ┃ ┊  
    ┊ ┃ ┃  ┏┻━┓ ┊   ┏━┻━┓ ┃ ┊  
0.44┊ ┃ ┃  6  ┃ ┊   6   ┃ ┃ ┊  
    ┊ ┃ ┃ ┏┻┓ ┃ ┊  ┏┻━┓ ┃ ┃ ┊  
0.42┊ ┃ ┃ ┃ ┃ ┃ ┊  5  ┃ ┃ ┃ ┊  
    ┊ ┃ ┃ ┃ ┃ ┃ ┊ ┏┻┓ ┃ ┃ ┃ ┊  
0.00┊ 0 1 2 3 4 ┊ 0 3 2 4 1 ┊  
  0.00        30.14       40.00

IbdSegment(pair=(0, 1), mrca=8, left=0, right=30.14286231249571)
IbdSegment(pair=(0, 2), mrca=9, left=0, right=30.14286231249571)
IbdSegment(pair=(0, 3), mrca=9, left=0, right=30.14286231249571)
IbdSegment(pair=(0, 4), mrca=9, left=0, right=30.14286231249571)
IbdSegment(pair=(0, 1), mrca=9, left=30.14286231249571, right=40.0)
IbdSegment(pair=(0, 2), mrca=6, left=30.14286231249571, right=40.0)
IbdSegment(pair=(0, 3), mrca=5, left=30.14286231249571, right=40.0)
IbdSegment(pair=(0, 4), mrca=7, left=30.14286231249571, right=40.0)
IbdSegment(pair=(1, 2), mrca=9, left=0, right=40.0)
IbdSegment(pair=(1, 3), mrca=9, left=0, right=40.0)
IbdSegment(pair=(1, 4), mrca=9, left=0, right=40.0)
IbdSegment(pair=(2, 3), mrca=6, left=0, right=40.0)
IbdSegment(pair=(2, 4), mrca=7, left=0, right=40.0)
IbdSegment(pair=(3, 4), mrca=7, left=0, right=40.0)
```

Is this what you were looking for? Note that this is quite a strong definition of IBD, in that the MRCAs must be identical, and this may not be what you want.

**Comment by RomainPechey** [comment]
Hello,

I am writing to you because I am looking for a method to respond to exactly the same problem as that described by hancockzb. The answer you provided responds perfectly, however this answer is a few years old now and I wanted to know if a more “optimal” method has been developed since. In particular, I saw the squash() function in the documentation, which seemed designed for this purpose, but I have not managed to use it for this purpose.

I would be grateful for your insight.

Thank you in advance. 

Romain PECHEY

**Reply by jeromekelleher** [reply]
Hi @RomainPechey :wave:

There is the [ibd_segments](https://tskit.dev/tskit/docs/stable/ibd.html) method in tskit now, which is much more efficient. However, it computes something slightly different, which many not be exactly what you want. See [this preprint](https://www.biorxiv.org/content/10.1101/2024.11.30.626138v1) for some discussion on the point.

**Reply by RomainPechey** [reply]
Thank you very much for your reply.

I must have been unclear in my request, I apologize. I was aware of the ibd_segments() function. However, when I use this function, the segments are not stitched together to form the “real” IBD segments per pair of individuals, unlike the method you presented in response to hancockzb. To be clearer, if I take the example in the discussion, with the ibd_segments() method, I don't get stitched segments such as:

IbdSegment(pair=(3, 4), mrca=7, left=0, right=40.0)
 
Instead, I get two segments:
 
IbdSegment(pair=(3, 4), mrca=7, left=0, right=30.14286231249571)
IbdSegment(pair=(3, 4), mrca=7, left=30.14286231249571, right=40.0)
 
This does not correspond to what I am looking for. Hence my question about the squash() function which, from what I understood, seemed to solve this problem, but this does not seem to be the case. As for the extend_haplotypes() function, I have also tried it, but it does not solve the problem I have outlined either.

I hope that my answer sheds enough light to identify whether a solution is possible.

Thank you again for your responsiveness.

Cordially,

Romain PECHEY

**Reply by jeromekelleher** [reply]
I see. I'm afraid the answer is "no" then I think. We haven't gotten back to the IBD segments to address these issues.

**Reply by RomainPechey** [reply]
Hello,

We are writing to you again regarding the previous issue, for which you indicated that you had not yet found a solution.

In collaboration with Iago Bonnici (@iago-lito) and Océane Eychenne (@OceaneEYCHENNE), we have developed a script that resolves the issue, namely the assembly of IBD segments between all pairs of sampled individuals.
Our script has the dual advantage of being fast to execute and robust in taking into account parameters that can be added to the ibd_segments() function.

Attached you will find the code and an explanatory document schematically presenting how the combination of the convolution function and the concatenation function, which is at the heart of our approach.

We would be delighted to have your feedback and comments.

Best, 

Iago Bonnici, Océane Eychenne, and Romain Pechey

[Convolution_and_concatenation.pdf](https://github.com/user-attachments/files/19845568/Convolution_and_concatenation.pdf)
[IBD_extract_and_stitch.txt](https://github.com/user-attachments/files/19845574/IBD_extract_and_stitch.txt)

**Reply by jeromekelleher** [reply]
This looks really nice,  thanks @RomainPechey! I'm a little worried that the code will be lost here - perhaps you could open a new "show-and-tell" discussion with the code and explanation inlined? I think this would make it much more accessible and likely to be found and used.

**Reply by iago-lito** [reply]
Hello @jeromekelleher, we weren't exactly sure what would be most useful for sharing. Are you suggesting that we'd dedicate another github "discussion" to this? Or a wiki section? Or would it make sense to have the snippet directly integrated as a tskit util?

**Reply by jeromekelleher** [reply]
Either another specific GitHub discussion (marked "show and tell", there's a type to choose when you start a new discussion) or just put code and explanation inline here. Attached files just aren't very visible.

**Reply by gregorgorjanc** [reply]
Tutorial?

**Reply by RomainPechey** [reply]
Thanks for your replies. I've created a new discussion in the “show and tell” section with the code and explanation inlined.
